File: apps/website/internal/handlers/cms.go
package handlers

import (
	"errors"
	"net/http"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/google/uuid"
	"gorm.io/gorm"

	"github.com/tbd-milsim/reforger-backend/internal/middleware"
	"github.com/tbd-milsim/reforger-backend/internal/models"
	"github.com/tbd-milsim/reforger-backend/internal/services"
)

const maxUploadBytes = 5 << 20 // 5MB

var allowedImageExt = map[string]bool{".jpg": true, ".jpeg": true, ".png": true, ".webp": true}

// validTag normalizes/validates an announcement category tag.
func validTag(s string) (models.AnnouncementTag, bool) {
	switch models.AnnouncementTag(s) {
	case models.TagUpdate, models.TagEvent, models.TagModpackUpdate, models.TagImportant:
		return models.AnnouncementTag(s), true
	case "":
		return models.TagUpdate, true
	default:
		return "", false
	}
}

// snippetFrom derives a preview if the author did not supply one.
func snippetFrom(explicit, body string) string {
	if explicit != "" {
		return explicit
	}
	return services.Snippet(body, 200)
}

// announcementInput is the body for creating an announcement.
type announcementInput struct {
	Title         string `json:"title" binding:"required"`
	Body          string `json:"body" binding:"required"`
	Snippet       string `json:"snippet"`
	Tag           string `json:"tag"`
	ThumbnailURL  string `json:"thumbnail_url"`
	IsPinned      bool   `json:"is_pinned"`
	Status        string `json:"status"` // draft (default) | published
	PushToDiscord bool   `json:"push_to_discord"`
}

// CreateAnnouncement creates a draft or published announcement and optionally
// pushes it to the Discord #announcements webhook.
//
// @route POST /api/v1/cms/announcements
func (h *Handler) CreateAnnouncement(c *gin.Context) {
	var in announcementInput
	if err := c.ShouldBindJSON(&in); err != nil {
		logHandlerErr(c, "CreateAnnouncement", http.StatusBadRequest, "title and body are required")
		c.JSON(http.StatusBadRequest, gin.H{"error": "title and body are required"})
		return
	}
	tag, ok := validTag(in.Tag)
	if !ok {
		logHandlerErr(c, "CreateAnnouncement", http.StatusBadRequest, "invalid tag")
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid tag"})
		return
	}
	author := middleware.DiscordID(c)

	// Sanitize author-supplied HTML before it ever touches the DB so the body is
	// safe to render verbatim on the client (no stored XSS).
	body := services.SanitizeHTML(in.Body)

	a := models.Announcement{
		Title:        in.Title,
		Body:         body,
		Snippet:      snippetFrom(in.Snippet, body),
		Tag:          tag,
		ThumbnailURL: in.ThumbnailURL,
		AuthorID:     author,
		IsPinned:     in.IsPinned,
		Status:       models.AnnouncementDraft,
	}
	if in.Status == string(models.AnnouncementPublished) {
		a.Status = models.AnnouncementPublished
		now := time.Now()
		a.PublishedAt = &now
	}

	if err := h.db.Create(&a).Error; err != nil {
		logHandlerErr(c, "CreateAnnouncement", http.StatusInternalServerError, "could not create announcement")
		c.JSON(http.StatusInternalServerError, gin.H{"error": "could not create announcement"})
		return
	}

	if a.Status == models.AnnouncementPublished && in.PushToDiscord {
		h.pushToDiscord(c, &a)
	}

	name := author
	//nolint:errcheck // best-effort: audit log is non-blocking; a failed write must not fail the request.
	_ = services.WriteAudit(h.db, models.SeverityInfo, &author, name,
		"announcement.create", name+" created announcement '"+a.Title+"'", "announcement", a.ID.String())

	c.JSON(http.StatusCreated, a)
}

// announcementUpdate is a partial update; only present fields are applied.
type announcementUpdate struct {
	Title         *string `json:"title"`
	Body          *string `json:"body"`
	Snippet       *string `json:"snippet"`
	Tag           *string `json:"tag"`
	ThumbnailURL  *string `json:"thumbnail_url"`
	IsPinned      *bool   `json:"is_pinned"`
	Status        *string `json:"status"`
	PushToDiscord *bool   `json:"push_to_discord"`
}

// UpdateAnnouncement edits an announcement, supporting draft->published
// transitions (which set published_at and may push to Discord).
//
// @route PATCH /api/v1/cms/announcements/:id
//
//nolint:cyclop // partial-update handler branches per optional field + draft->published transition; splitting tracked SIZE-3 debt (T-125.4).
func (h *Handler) UpdateAnnouncement(c *gin.Context) {
	id, err := uuid.Parse(c.Param("id"))
	if err != nil {
		logHandlerErr(c, "UpdateAnnouncement", http.StatusBadRequest, "invalid id")
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid id"})
		return
	}
	var a models.Announcement
	if err := h.db.First(&a, "id = ?", id).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "announcement not found"})
		return
	}

	var in announcementUpdate
	if err := c.ShouldBindJSON(&in); err != nil {
		logHandlerErr(c, "UpdateAnnouncement", http.StatusBadRequest, "invalid body")
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid body"})
		return
	}

	updates := map[string]any{}
	if in.Title != nil {
		updates["title"] = *in.Title
	}
	if in.Body != nil {
		updates["body"] = services.SanitizeHTML(*in.Body)
	}
	if in.Snippet != nil {
		updates["snippet"] = *in.Snippet
	}
	if in.Tag != nil {
		tag, ok := validTag(*in.Tag)
		if !ok {
			logHandlerErr(c, "UpdateAnnouncement", http.StatusBadRequest, "invalid tag")
			c.JSON(http.StatusBadRequest, gin.H{"error": "invalid tag"})
			return
		}
		updates["tag"] = tag
	}
	if in.ThumbnailURL != nil {
		updates["thumbnail_url"] = *in.ThumbnailURL
	}
	if in.IsPinned != nil {
		updates["is_pinned"] = *in.IsPinned
	}

	nowPublishing := false
	if in.Status != nil {
		switch models.AnnouncementStatus(*in.Status) {
		case models.AnnouncementDraft, models.AnnouncementPublished, models.AnnouncementArchived:
			updates["status"] = *in.Status
			if models.AnnouncementStatus(*in.Status) == models.AnnouncementPublished && a.PublishedAt == nil {
				now := time.Now()
				updates["published_at"] = now
				nowPublishing = true
			}
		default:
			logHandlerErr(c, "UpdateAnnouncement", http.StatusBadRequest, "invalid status")
			c.JSON(http.StatusBadRequest, gin.H{"error": "invalid status"})
			return
		}
	}

	if len(updates) > 0 {
		if err := h.db.Model(&a).Updates(updates).Error; err != nil {
			logHandlerErr(c, "UpdateAnnouncement", http.StatusInternalServerError, "could not update announcement")
			c.JSON(http.StatusInternalServerError, gin.H{"error": "could not update announcement"})
			return
		}
	}
	if err := h.db.First(&a, "id = ?", id).Error; err != nil {
		logHandlerErr(c, "UpdateAnnouncement", http.StatusInternalServerError, "reload after update failed")
		c.JSON(http.StatusInternalServerError, gin.H{"error": "could not load announcement"})
		return
	}

	if in.PushToDiscord != nil && *in.PushToDiscord &&
		a.Status == models.AnnouncementPublished && (nowPublishing || !a.PushedToDiscord) {
		h.pushToDiscord(c, &a)
		if err := h.db.First(&a, "id = ?", id).Error; err != nil && !errors.Is(err, gorm.ErrRecordNotFound) {
			logHandlerErr(c, "UpdateAnnouncement", http.StatusOK, "announcement re-reload failed")
		}
	}

	c.JSON(http.StatusOK, a)
}

// DeleteAnnouncement archives an announcement (removes it from the feed but
// keeps it recoverable).
//
// @route DELETE /api/v1/cms/announcements/:id
func (h *Handler) DeleteAnnouncement(c *gin.Context) {
	id, err := uuid.Parse(c.Param("id"))
	if err != nil {
		logHandlerErr(c, "DeleteAnnouncement", http.StatusBadRequest, "invalid id")
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid id"})
		return
	}
	res := h.db.Model(&models.Announcement{}).Where("id = ?", id).
		Update("status", models.AnnouncementArchived)
	if res.Error != nil {
		logHandlerErr(c, "DeleteAnnouncement", http.StatusInternalServerError, "could not delete announcement")
		c.JSON(http.StatusInternalServerError, gin.H{"error": "could not delete announcement"})
		return
	}
	if res.RowsAffected == 0 {
		c.JSON(http.StatusNotFound, gin.H{"error": "announcement not found"})
		return
	}
	c.Status(http.StatusNoContent)
}

// PushAnnouncementDiscord manually (re)pushes an announcement to the webhook.
//
// @route POST /api/v1/cms/announcements/:id/push-discord
func (h *Handler) PushAnnouncementDiscord(c *gin.Context) {
	id, err := uuid.Parse(c.Param("id"))
	if err != nil {
		logHandlerErr(c, "PushAnnouncementDiscord", http.StatusBadRequest, "invalid id")
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid id"})
		return
	}
	if !h.webhook.Enabled() {
		logHandlerErr(c, "PushAnnouncementDiscord", http.StatusBadRequest, "discord webhook not configured")
		c.JSON(http.StatusBadRequest, gin.H{"error": "discord webhook not configured"})
		return
	}
	var a models.Announcement
	if err := h.db.First(&a, "id = ?", id).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "announcement not found"})
		return
	}
	if ok := h.pushToDiscord(c, &a); !ok {
		logHandlerErr(c, "PushAnnouncementDiscord", http.StatusBadGateway, "webhook push failed")
		c.JSON(http.StatusBadGateway, gin.H{"error": "webhook push failed"})
		return
	}
	c.JSON(http.StatusOK, gin.H{"pushed": true})
}

// pushToDiscord pushes an announcement and records the result. On failure it
// writes a CRIT audit row and returns false. On success it persists the message
// id and pushed flag.
func (h *Handler) pushToDiscord(c *gin.Context, a *models.Announcement) bool {
	msgID, err := h.webhook.PushAnnouncement(c.Request.Context(), a)
	if err != nil {
		//nolint:errcheck // best-effort: audit log is non-blocking; a failed write must not fail the request.
		_ = services.WriteAudit(h.db, models.SeverityCrit, nil, "system",
			"webhook.push_failed",
			"Webhook failed to push payload to Discord channel #announcements ('"+a.Title+"')",
			"announcement", a.ID.String())
		return false
	}
	h.db.Model(&models.Announcement{}).Where("id = ?", a.ID).
		Updates(map[string]any{"pushed_to_discord": true, "discord_message_id": msgID})
	a.PushedToDiscord = true
	a.DiscordMessageID = msgID
	return true
}

// UploadImage accepts a thumbnail (multipart "file") and returns its URL.
// Local-disk storage for now; swap for S3/MinIO in production.
//
// @route POST /api/v1/cms/uploads
func (h *Handler) UploadImage(c *gin.Context) {
	file, err := c.FormFile("file")
	if err != nil {
		logHandlerErr(c, "UploadImage", http.StatusBadRequest, "file field required")
		c.JSON(http.StatusBadRequest, gin.H{"error": "file field required"})
		return
	}
	if file.Size > maxUploadBytes {
		logHandlerErr(c, "UploadImage", http.StatusRequestEntityTooLarge, "file exceeds 5MB")
		c.JSON(http.StatusRequestEntityTooLarge, gin.H{"error": "file exceeds 5MB"})
		return
	}
	ext := strings.ToLower(filepath.Ext(file.Filename))
	if !allowedImageExt[ext] {
		c.JSON(http.StatusUnsupportedMediaType, gin.H{"error": "only JPG, PNG, WEBP allowed"})
		return
	}
	if err := os.MkdirAll(uploadDir, 0o755); err != nil {
		logHandlerErr(c, "UploadImage", http.StatusInternalServerError, "storage unavailable")
		c.JSON(http.StatusInternalServerError, gin.H{"error": "storage unavailable"})
		return
	}
	name := uuid.NewString() + ext
	if err := c.SaveUploadedFile(file, filepath.Join(uploadDir, name)); err != nil {
		logHandlerErr(c, "UploadImage", http.StatusInternalServerError, "could not save file")
		c.JSON(http.StatusInternalServerError, gin.H{"error": "could not save file"})
		return
	}
	c.JSON(http.StatusCreated, gin.H{"url": "/uploads/" + name})
}

File: apps/website/src/handlers/cms.rs
//! CMS — announcements CRUD + Discord push + image upload. Rust port of `handlers/cms.go`.

use axum::extract::rejection::JsonRejection;
use axum::extract::{Multipart, Path, State};
use axum::http::StatusCode;
use axum::response::Json;
use serde::Deserialize;
use serde_json::{Value, json};
use uuid::Uuid;

use crate::error::ApiError;
use crate::handlers::field_tools::UPLOAD_DIR;
use crate::handlers::username;
use crate::middleware::AdminUser;
use crate::models::{Announcement, AnnouncementStatus, AnnouncementTag, AuditSeverity};
use crate::services::{sanitize_html, snippet, write_audit};
use crate::state::AppState;

const MAX_UPLOAD_BYTES: usize = 5 << 20;

fn valid_tag(s: &str) -> Option<AnnouncementTag> {
    match s {
        "" | "update" => Some(AnnouncementTag::Update),
        "event" => Some(AnnouncementTag::Event),
        "modpack_update" => Some(AnnouncementTag::ModpackUpdate),
        "important" => Some(AnnouncementTag::Important),
        _ => None,
    }
}

fn snippet_from(explicit: &str, body: &str) -> String {
    if !explicit.is_empty() {
        explicit.to_string()
    } else {
        snippet(body, 200)
    }
}

/// Push an announcement to the webhook; record the result. Returns success.
async fn push_to_discord(state: &AppState, a: &Announcement) -> bool {
    match state.webhook.push_announcement(a).await {
        Ok(msg_id) => {
            let _ = sqlx::query(
                "UPDATE announcements SET pushed_to_discord = true, discord_message_id = $1 WHERE id = $2",
            )
            .bind(&msg_id)
            .bind(a.id)
            .execute(&state.pool)
            .await;
            true
        }
        Err(_) => {
            write_audit(
                &state.pool,
                AuditSeverity::Crit,
                None,
                "system",
                "webhook.push_failed",
                &format!(
                    "Webhook failed to push payload to Discord channel #announcements ('{}')",
                    a.title
                ),
                "announcement",
                &a.id.to_string(),
            )
            .await;
            false
        }
    }
}

#[derive(Debug, Deserialize)]
pub struct AnnouncementInput {
    #[serde(default)]
    title: String,
    #[serde(default)]
    body: String,
    #[serde(default)]
    snippet: String,
    #[serde(default)]
    tag: String,
    #[serde(default)]
    thumbnail_url: String,
    #[serde(default)]
    is_pinned: bool,
    #[serde(default)]
    status: String,
    #[serde(default)]
    push_to_discord: bool,
}

/// `POST /api/v1/cms/announcements` — create draft/published (+ optional push).
///
/// @route POST /api/v1/cms/announcements
pub async fn create_announcement(
    State(state): State<AppState>,
    admin: AdminUser,
    body: Result<Json<AnnouncementInput>, JsonRejection>,
) -> Result<(StatusCode, Json<Announcement>), ApiError> {
    let Json(input) = body.map_err(|_| ApiError::bad_request("title and body are required"))?;
    if input.title.is_empty() || input.body.is_empty() {
        return Err(ApiError::bad_request("title and body are required"));
    }
    let Some(tag) = valid_tag(&input.tag) else {
        return Err(ApiError::bad_request("invalid tag"));
    };
    let author = &admin.0.discord_id;
    // Sanitize author-supplied HTML before persist (no stored XSS).
    let body_html = sanitize_html(&input.body);
    let snip = snippet_from(&input.snippet, &body_html);
    let published = input.status == "published";

    let a: Announcement = sqlx::query_as(
        "INSERT INTO announcements \
         (title, body, snippet, tag, thumbnail_url, author_id, is_pinned, status, published_at, \
          pushed_to_discord, discord_message_id, created_at, updated_at) \
         VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, false, '', now(), now()) RETURNING id, title, body, COALESCE(snippet, '') AS snippet, tag, COALESCE(thumbnail_url, '') AS thumbnail_url, author_id, status, is_pinned, pushed_to_discord, COALESCE(discord_message_id, '') AS discord_message_id, published_at, COALESCE(created_at, '0001-01-01 00:00:00+00'::timestamptz) AS created_at, COALESCE(updated_at, '0001-01-01 00:00:00+00'::timestamptz) AS updated_at",
    )
    .bind(&input.title)
    .bind(&body_html)
    .bind(&snip)
    .bind(tag)
    .bind(&input.thumbnail_url)
    .bind(author)
    .bind(input.is_pinned)
    .bind(if published {
        AnnouncementStatus::Published
    } else {
        AnnouncementStatus::Draft
    })
    .bind(if published {
        Some(chrono::Utc::now())
    } else {
        None
    })
    .fetch_one(&state.pool)
    .await?;

    let mut a = a;
    if published && input.push_to_discord {
        push_to_discord(&state, &a).await;
        a = reload(&state, a.id).await?.unwrap_or(a);
    }
    let name = username(&state.pool, author).await;
    write_audit(
        &state.pool,
        AuditSeverity::Info,
        Some(author),
        &name,
        "announcement.create",
        &format!("{name} created announcement '{}'", a.title),
        "announcement",
        &a.id.to_string(),
    )
    .await;
    Ok((StatusCode::CREATED, Json(a)))
}

#[derive(Debug, Deserialize)]
pub struct AnnouncementUpdate {
    title: Option<String>,
    body: Option<String>,
    snippet: Option<String>,
    tag: Option<String>,
    thumbnail_url: Option<String>,
    is_pinned: Option<bool>,
    status: Option<String>,
    push_to_discord: Option<bool>,
}

/// `PATCH /api/v1/cms/announcements/:id` — partial edit (+ draft→published push).
///
/// @route PATCH /api/v1/cms/announcements/:id
pub async fn update_announcement(
    State(state): State<AppState>,
    _a: AdminUser,
    Path(id): Path<String>,
    body: Result<Json<AnnouncementUpdate>, JsonRejection>,
) -> Result<Json<Announcement>, ApiError> {
    let Ok(id) = Uuid::parse_str(&id) else {
        return Err(ApiError::bad_request("invalid id"));
    };
    let Some(existing) = reload(&state, id).await? else {
        return Err(ApiError::not_found("announcement not found"));
    };
    let Json(input) = body.map_err(|_| ApiError::bad_request("invalid body"))?;

    let mut qb: sqlx::QueryBuilder<sqlx::Postgres> =
        sqlx::QueryBuilder::new("UPDATE announcements SET updated_at = now()");
    if let Some(t) = &input.title {
        qb.push(", title = ").push_bind(t.clone());
    }
    if let Some(b) = &input.body {
        qb.push(", body = ").push_bind(sanitize_html(b));
    }
    if let Some(s) = &input.snippet {
        qb.push(", snippet = ").push_bind(s.clone());
    }
    if let Some(t) = &input.tag {
        let Some(tag) = valid_tag(t) else {
            return Err(ApiError::bad_request("invalid tag"));
        };
        qb.push(", tag = ").push_bind(tag);
    }
    if let Some(u) = &input.thumbnail_url {
        qb.push(", thumbnail_url = ").push_bind(u.clone());
    }
    if let Some(p) = input.is_pinned {
        qb.push(", is_pinned = ").push_bind(p);
    }
    let mut now_publishing = false;
    if let Some(s) = &input.status {
        let status = match s.as_str() {
            "draft" => AnnouncementStatus::Draft,
            "published" => AnnouncementStatus::Published,
            "archived" => AnnouncementStatus::Archived,
            _ => return Err(ApiError::bad_request("invalid status")),
        };
        qb.push(", status = ").push_bind(status);
        if status == AnnouncementStatus::Published && existing.published_at.is_none() {
            qb.push(", published_at = now()");
            now_publishing = true;
        }
    }
    qb.push(" WHERE id = ").push_bind(id);
    qb.build()
        .execute(&state.pool)
        .await
        .map_err(ApiError::from)?;

    let mut a = reload(&state, id)
        .await?
        .ok_or_else(|| ApiError::internal("could not load announcement"))?;
    if input.push_to_discord == Some(true)
        && a.status == AnnouncementStatus::Published
        && (now_publishing || !a.pushed_to_discord)
    {
        push_to_discord(&state, &a).await;
        a = reload(&state, id).await?.unwrap_or(a);
    }
    Ok(Json(a))
}

/// `DELETE /api/v1/cms/announcements/:id` — archive (recoverable).
///
/// @route DELETE /api/v1/cms/announcements/:id
pub async fn delete_announcement(
    State(state): State<AppState>,
    _a: AdminUser,
    Path(id): Path<String>,
) -> Result<StatusCode, ApiError> {
    let Ok(id) = Uuid::parse_str(&id) else {
        return Err(ApiError::bad_request("invalid id"));
    };
    let res = sqlx::query("UPDATE announcements SET status = 'archived' WHERE id = $1")
        .bind(id)
        .execute(&state.pool)
        .await?;
    if res.rows_affected() == 0 {
        return Err(ApiError::not_found("announcement not found"));
    }
    Ok(StatusCode::NO_CONTENT)
}

/// `POST /api/v1/cms/announcements/:id/push-discord` — manual (re)push.
///
/// @route POST /api/v1/cms/announcements/:id/push-discord
pub async fn push_announcement_discord(
    State(state): State<AppState>,
    _a: AdminUser,
    Path(id): Path<String>,
) -> Result<Json<Value>, ApiError> {
    let Ok(id) = Uuid::parse_str(&id) else {
        return Err(ApiError::bad_request("invalid id"));
    };
    if !state.webhook.enabled() {
        return Err(ApiError::bad_request("discord webhook not configured"));
    }
    let Some(a) = reload(&state, id).await? else {
        return Err(ApiError::not_found("announcement not found"));
    };
    if !push_to_discord(&state, &a).await {
        return Err(ApiError::new(
            StatusCode::BAD_GATEWAY,
            "webhook push failed",
        ));
    }
    Ok(Json(json!({ "pushed": true })))
}

/// `POST /api/v1/cms/uploads` — thumbnail upload (multipart "file").
///
/// @route POST /api/v1/cms/uploads
pub async fn upload_image(
    _a: AdminUser,
    mut mp: Multipart,
) -> Result<(StatusCode, Json<Value>), ApiError> {
    while let Some(field) = mp
        .next_field()
        .await
        .map_err(|_| ApiError::bad_request("file field required"))?
    {
        if field.name() != Some("file") {
            continue;
        }
        let filename = field.file_name().unwrap_or("").to_string();
        let ext = ext_lower(&filename);
        let data = field
            .bytes()
            .await
            .map_err(|_| ApiError::bad_request("file field required"))?;
        if data.len() > MAX_UPLOAD_BYTES {
            return Err(ApiError::new(
                StatusCode::PAYLOAD_TOO_LARGE,
                "file exceeds 5MB",
            ));
        }
        if !matches!(ext.as_str(), ".jpg" | ".jpeg" | ".png" | ".webp") {
            return Err(ApiError::new(
                StatusCode::UNSUPPORTED_MEDIA_TYPE,
                "only JPG, PNG, WEBP allowed",
            ));
        }
        std::fs::create_dir_all(UPLOAD_DIR)
            .map_err(|_| ApiError::internal("storage unavailable"))?;
        let name = format!("{}{ext}", Uuid::new_v4());
        std::fs::write(format!("{UPLOAD_DIR}/{name}"), &data)
            .map_err(|_| ApiError::internal("could not save file"))?;
        return Ok((
            StatusCode::CREATED,
            Json(json!({ "url": format!("/uploads/{name}") })),
        ));
    }
    Err(ApiError::bad_request("file field required"))
}

/// Lowercase file extension including the dot (`.jpg`), or empty.
fn ext_lower(filename: &str) -> String {
    match filename.rsplit_once('.') {
        Some((_, ext)) if !ext.is_empty() => format!(".{}", ext.to_lowercase()),
        _ => String::new(),
    }
}

/// Load one announcement by id (no soft-delete filter — matches Go's `First` on a
/// model without `DeletedAt`… announcements are archived, not soft-deleted here).
async fn reload(state: &AppState, id: Uuid) -> Result<Option<Announcement>, ApiError> {
    sqlx::query_as("SELECT id, title, body, COALESCE(snippet, '') AS snippet, tag, COALESCE(thumbnail_url, '') AS thumbnail_url, author_id, status, is_pinned, pushed_to_discord, COALESCE(discord_message_id, '') AS discord_message_id, published_at, COALESCE(created_at, '0001-01-01 00:00:00+00'::timestamptz) AS created_at, COALESCE(updated_at, '0001-01-01 00:00:00+00'::timestamptz) AS updated_at FROM announcements WHERE id = $1 AND deleted_at IS NULL")
        .bind(id)
        .fetch_optional(&state.pool)
        .await
        .map_err(ApiError::from)
}
